### PR TITLE
CSV export for planning

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: inc/planningcsv.class.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    */
+   public function output($text = false) {
+      $this->getlines();
+
+      if ($text === false) {
+         header("Expires: Mon, 26 Nov 1962 00:00:00 GMT");
+         header('Pragma: private'); /// IE BUG + SSL
+         header('Cache-control: private, must-revalidate'); /// IE BUG + SSL
+         header("Content-disposition: filename=".$this->filename);
+         header('Content-type: application/octetstream');
```
Comment (orthagh) [inline inc/planningcsv.class.php]:
```suggestion
         header('Content-type: text/csv');
```

File: inc/planningcsv.class.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+         // zero width no break space (for excel)
+      }
+
+      echo"\xEF\xBB\xBF";
```
Comment (orthagh) [inline inc/planningcsv.class.php]: maybe add a constant for bom

File: inc/planningcsv.class.php [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            $dateBegin = new DateTime($val["begin"]);
+            $dateBegin->setTimeZone(new DateTimeZone('UTC'));
+
+            $dateEnd = new DateTime($val["end"]);
+            $dateEnd->setTimeZone(new DateTimeZone('UTC'));
+
+            $itemtype = new $val['itemtype'];
+            //(acteur;titre item;id item;date-heure début,date-heure fin;catégorie)
+            $this->lines[] = [
+               'actor'     => $val['users_id'],
```
Comment (cedric-anne) [inline inc/planningcsv.class.php]: I would export name of user using `User::getRawName()` or something human readable..

File: inc/planningcsv.class.php [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+               'title'     => $val['name'],
+               'itemtype'  => $itemtype->getType(),
```
Comment (cedric-anne) [inline inc/planningcsv.class.php]: I would export name of item based on `$itemtype->getTypeName(1);`.

File: inc/planningcsv.class.php [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+               'items_id'  => $val[$itemtype->getForeignKeyField()],
```
Comment (cedric-anne) [inline inc/planningcsv.class.php]: Having the name of the item could be usefull.

File: inc/planningcsv.class.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+**/
+class PlanningCsv extends CommonGLPI {
+
+   private $users_id;
+   private $groups_id;
+   private $limititemtype;
+
+   private $eol = "\r\n";
+   private $quote = '"';
+   private $titles = [
```
Comment (cedric-anne) [inline inc/planningcsv.class.php]: I would use translated strings for CSV header.

File: inc/planningcsv.class.php (hunk 5 above)
Comment (trasher) [reply inc/planningcsv.class.php]: Exept for category (I did not find it or I was unsure) I've added the fields that were listed in the rfc :)

File: inc/planningcsv.class.php (hunk 2 above)
Comment (trasher) [reply inc/planningcsv.class.php]: This was copy/pasted from Search; but I'm not sure this is really required, I've dropped for now

File: inc/planning.class.php [hunk 7]
```diff
@@ -931,6 +931,11 @@ class='".$filter_data['type']."'>";
             echo "<li><a target='_blank' href='webcal://".$url['host'].":$port".
                  (isset($url['path'])?$url['path']:'')."$cal_url'>".
                  _sx("button", "Export")." - ".__("Webcal")."</a></li>";
+
+            echo "<li><a target='_blank' href='".$CFG_GLPI['root_doc'].
+                 "/front/planningcsv.php?uID=".$loginUser->fields['id']."'>".
```
Comment (cedric-anne) [inline inc/planning.class.php]: uID/gID params should correspond to selected user/group
```suggestion
                 "/front/planningcsv.php?uID=".$uID."&gID=".$gID."'>".
```

File: inc/planningcsv.class.php [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            $dateBegin = new DateTime($val["begin"]);
+            $dateBegin->setTimeZone(new DateTimeZone('UTC'));
+
+            $dateEnd = new DateTime($val["end"]);
+            $dateEnd->setTimeZone(new DateTimeZone('UTC'));
+
+            $itemtype = new $val['itemtype'];
+
+            $user = new User();
+            $user->getFromDB($val['users_id']);
```
Comment (cedric-anne) [inline inc/planningcsv.class.php]: What do we want to be exported for `actor` field ? `users_id` field value correspond to the author of the item, not the assigned user.

- For `CommonITILTask`, assigned user is in `users_id_tech` field,
- For `ProjectTask` and `Reminder`, target users are not available in items returned by `populatePlanning`.

File: front/planningcsv.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+ */
+
+include ('../inc/includes.php');
+
+if (!isset($_GET['genical'])) {
+   Session::checkRight("planning", READ);
+}
+
+$users_id = null;
+$groups_id = (isset($_GET["gID"]) ? (int)$_GET['uID'] : 0);
```
[the remark this answers is not in the document]
Comment (cedric-anne) [reply front/planningcsv.php]: Still not fixed.

File: inc/planningcsv.class.php (hunk 8 above)
Comment (trasher) [reply inc/planningcsv.class.php]: ping @orthagh

**Comment by cedric-anne** [comment]
I also had an SQL error when trying to export planning for a group:
```
  *** MySQL query error:
  SQL: SELECT DISTINCT `glpi_reminders`.* FROM `glpi_reminders` LEFT JOIN `glpi_reminders_users` ON (`glpi_reminders_users`.`reminders_id` = `glpi_reminders`.`id`) LEFT JOIN `glpi_groups_reminders` ON (`glpi_groups_reminders`.`reminders_id` = `glpi_reminders`.`id`) LEFT JOIN `glpi_profiles_reminders` ON (`glpi_profiles_reminders`.`reminders_id` = `glpi_reminders`.`id`) LEFT JOIN `glpi_entities_reminders` ON (`glpi_entities_reminders`.`reminders_id` = `glpi_reminders`.`id`) WHERE `glpi_reminders`.`is_planned` = '1' AND `begin` < '2020-09-17 09:26:24' AND `end` > '2018-09-28 09:26:24' AND `glpi_reminders`.`groups_id` = '0' ORDER BY `begin`
  Error: Unknown column 'glpi_reminders.groups_id' in 'where clause'
  Backtrace :
  inc/dbmysqliterator.class.php:95                   
  inc/dbmysql.class.php:851                          DBmysqlIterator->execute()
  inc/reminder.class.php:984                         DBmysql->request()
  inc/planningcsv.class.php:96                       Reminder::populatePlanning()
  inc/planningcsv.class.php:164                      PlanningCsv->buildLines()
  inc/planningcsv.class.php:134                      PlanningCsv->getlines()
  front/planningcsv.php:108                          PlanningCsv->output()
```
Looks like it is and existing issue not related to this PR.

File: inc/planningcsv.class.php (hunk 8 above)
Comment (orthagh) [reply inc/planningcsv.class.php]: No i think it's ok, in the current case, users_id came from populatePlanning (in $val var).

**Pull request merged**